offset_manager.go

```diff
@@ -74,52 +86,292 @@ func (om *offsetManager) ManagePartition(topic string, partition int32) (Partiti
+		for retries := om.conf.Metadata.Retry.Max; true; {
+			if om.flushToBroker() {
+				break
+			}
+			if retries--; retries < 0 {
+				break
+			}
+		}
+
+		om.releasePOMs(true)
+		om.brokerMu.Lock()
+		om.broker = nil
+		om.brokerMu.Unlock()
+	})
 	return nil
 }
 
-func (om *offsetManager) refBrokerOffsetManager(broker *Broker) *brokerOffsetManager {
-	om.lock.Lock()
-	defer om.lock.Unlock()
+func (om *offsetManager) fetchInitialOffset(topic string, partition int32, retries int) (int64, string, error) {
+	broker, err := om.coordinator()
+	if err != nil {
+		if retries <= 0 {
+			return 0, "", err
+		}
+		return om.fetchInitialOffset(topic, partition, retries-1)
+	}
+
+	request := new(OffsetFetchRequest)
+	request.Version = 1
+	request.ConsumerGroup = om.group
+	request.AddPartition(topic, partition)
 
-	bom := om.boms[broker]
-	if bom == nil {
-		bom = om.newBrokerOffsetManager(broker)
-		om.boms[broker] = bom
+	response, err := broker.FetchOffset(request)
+	if err != nil {
+		if retries <= 0 {
+			return 0, "", err
+		}
+		om.releaseCoordinator(broker)
+		return om.fetchInitialOffset(topic, partition, retries-1)
 	}
 
-	bom.refs++
+	block := response.GetBlock(topic, partition)
+	if block == nil {
+		return 0, "", ErrIncompleteResponse
+	}
 
-	return bom
+	switch block.Err {
+	case ErrNoError:
+		return block.Offset, block.Metadata, nil
+	case ErrNotCoordinatorForConsumer:
+		if retries <= 0 {
+			return 0, "", block.Err
+		}
+		om.releaseCoordinator(broker)
+		return om.fetchInitialOffset(topic, partition, retries-1)
+	case ErrOffsetsLoadInProgress:
+		if retries <= 0 {
+			return 0, "", block.Err
+		}
+		select {
+		case <-om.closing:
+			return 0, "", block.Err
+		case <-time.After(om.conf.Metadata.Retry.Backoff):
+		}
+		return om.fetchInitialOffset(topic, partition, retries-1)
+	default:
+		return 0, "", block.Err
+	}
 }
 
-func (om *offsetManager) unrefBrokerOffsetManager(bom *brokerOffsetManager) {
-	om.lock.Lock()
-	defer om.lock.Unlock()
+func (om *offsetManager) coordinator() (*Broker, error) {
+	om.brokerMu.RLock()
+	broker := om.broker
+	om.brokerMu.RUnlock()
+
+	if broker != nil {
+		return broker, nil
+	}
 
-	bom.refs--
+	om.brokerMu.Lock()
+	defer om.brokerMu.Unlock()
 
-	if bom.refs == 0 {
-		close(bom.updateSubscriptions)
-		if om.boms[bom.broker] == bom {
-			delete(om.boms, bom.broker)
+	if broker := om.broker; broker != nil {
+		return broker, nil
+	}
+
+	if err := om.client.RefreshCoordinator(om.group); err != nil {
+		return nil, err
+	}
+
+	broker, err := om.client.Coordinator(om.group)
+	if err != nil {
+		return nil, err
+	}
+
+	om.broker = broker
+	return broker, nil
+}
+
+func (om *offsetManager) releaseCoordinator(b *Broker) {
+	om.brokerMu.Lock()
+	if om.broker == b {
+		om.broker = nil
+	}
+	om.brokerMu.Unlock()
+}
+
+func (om *offsetManager) mainLoop() {
+	defer om.ticker.Stop()
+	defer close(om.closed)
+
+	for {
+		select {
+		case <-om.ticker.C:
+			om.flushToBroker()
+			om.releasePOMs(false)
```

Comment:
It took me a while to figure out what this was doing, and I'm not much a fan of this pattern. An `updateSubscriptions` channel (as in the old BOM) seems easier to follow and less likely to result in races/deadlocks than polling a piece of state in all the POMs?

Reply:
I thought about `updateSubscriptions` but it's not making things any more trivial in my opinion. The problem is that the eviction of POMs is closely linked to the flush cycles and we still need to coordinate those on `OffsetManager.Close`, i.e. keep retrying flush until either `Retry.Max` is reached or until all POMs are clean. I don't see how to implement this with an `updateSubscriptions` channel, we would have to do something like:

```
func (om *offsetManager) Close() error {
  om.closeOnce.Do(func() {
    // exit the mainLoop
    close(om.closing)
    <-om.closed

    // mark all POMs as closed
    om.asyncClosePOMs()

    deadline := time.NewTimer(10*time.Second) // just an example
    defer deadline.Stop()

    retry := time.NewTicker(100*time.Millisecond) // just an example
    defer retry.Stop()

  FlushLoop:
    for {
      om.flushToBroker()

      select {
      case <-deadline.C:
        break FlushLoop
      case pom := <-om.updateSubscription:
        om.pomsLock.Lock()
        delete(om.poms[pom.topic], pom.partition)
        if len(om.poms[pom.topic]) == 0 {
          delete(om.poms, pom.topic)
        }
        remaining := len(om.poms)
        om.pomsLock.Unlock()
        if remaining == 0 {
          break FlushLoop
        }
      case <-retry.C:
      }
    }

    // TODO: abandon any remaining POMs

    om.brokerLock.Lock()
    om.broker = nil
    om.brokerLock.Unlock()
  })
  return nil
}
```

Even with helper method, I don't think this is easier to follow than the polling.

Reply:
For `OffsetManager.Close` I think you're overthinking things. Instead of exiting the mainloop and them manually looping flush, just call `pom.AsyncClose()` on all of them (as you already do), set a flag and then wait for a waitgroup. The mainloop can respect the flag to count down the number of remaining flushes.

Reply:
Letting `mainLoop` doing the work after calling `Close()` could delay things significantly, depending on `CommitInterval`. I'm not a fan of this, I prefer exiting the `mainLoop` to attempt an immediate final flush (with instant retries). The current approach will also simplify the integration with `ClusterClient`.

Reply:
OK. It's not clear to me how this will simplify the `ClusterClient` integration, but I don't feel too strongly about this.